hbase-client/src/main/java/org/apache/hadoop/hbase/client/MetricsConnection.java

```diff
@@ -47,13 +51,46 @@
 /**
  * This class is for maintaining the various connection statistics and publishing them through the
  * metrics interfaces. This class manages its own {@link MetricRegistry} and {@link JmxReporter} so
- * as to not conflict with other uses of Yammer Metrics within the client application. Instantiating
- * this class implicitly creates and "starts" instances of these classes; be sure to call
- * {@link #shutdown()} to terminate the thread pools they allocate.
+ * as to not conflict with other uses of Yammer Metrics within the client application. Calling
+ * {@link #getMetricsConnection()} implicitly creates and "starts" instances of these classes; be
+ * sure to call {@link #deleteMetricsConnection()} to terminate the thread pools they allocate. The
+ * metrics reporter will be shutdown {@link #shutdown()} when all connections within this metrics
+ * instances are closed.
  */
 @InterfaceAudience.Private
 public class MetricsConnection implements StatisticTrackable {
 
+  static final Map<String, MetricsConnection> METRICS_INSTANCES =
+    new HashMap<String, MetricsConnection>();
+
+  static MetricsConnection getMetricsConnection(final String scope,
+    Supplier<ThreadPoolExecutor> batchPool, Supplier<ThreadPoolExecutor> metaPool) {
+    MetricsConnection metrics;
+    synchronized (METRICS_INSTANCES) {
+      metrics = METRICS_INSTANCES.get(scope);
+      if (metrics == null) {
+        metrics = new MetricsConnection(scope, batchPool, metaPool);
+        METRICS_INSTANCES.put(scope, metrics);
+      } else {
+        metrics.addThreadPools(batchPool, metaPool);
+      }
+      metrics.incrConnectionCount();
+    }
+    return metrics;
+  }
+
+  static void deleteMetricsConnection(final String scope) {
+    synchronized (METRICS_INSTANCES) {
```

Comment:
The scope of this code block is not just per the single async connection object, it is also per the single metrics object which might be shared among multiple async connection objects.

Reply:
Is `String scope` here going to be unique per `AsyncConnectionImpl` object?

Reply:
If we use `hbase.client.metrics.scope`, then this `scope` value would not be unique across `AsyncConnectionImpl` objects IIUC.

Reply:
No, it might not be unique among connections. So multiple connection might share a common metrics object.

Reply:
Got it, regardless though, we can still avoid having to lock the entire map.